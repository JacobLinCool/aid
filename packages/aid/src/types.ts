export type PromiseOr<T> = T | Promise<T>;

export type LLMQuery = (
	messages: {
		role: "system" | "user" | "assistant";
		content: string;
	}[],
) => PromiseOr<string>;

/**
 * Options for an custom Aid Task.
 */
export interface AidTaskOptions<In extends string, Out> {
	/**
	 * The few-shot prompt examples.
	 */
	examples?: [string, Out][];
	/**
	 * The output schema strategy.
	 * @default "json-schema"
	 */
	strategy?: "ts" | "json-schema";
	/**
	 * Whether to check the output against the schema.
	 * @default false
	 */
	check?: boolean;
}
