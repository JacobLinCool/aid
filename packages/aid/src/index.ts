export * from "./aid";
export * from "./types";
