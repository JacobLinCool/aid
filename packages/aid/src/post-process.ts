// post process the output from the large language model, and make it parseable by the JSON parser
export function post_process(llm_output: string): string {
	llm_output = llm_output.trim();

	// remove possible markdown code fence at the beginning and end
	llm_output = llm_output.replace(/^```.*\n/m, "").replace(/```\n?$/m, "");

	return llm_output;
}
