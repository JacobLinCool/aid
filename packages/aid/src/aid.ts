import debug from "debug";
import type { OpenAI } from "openai";
import type { z } from "zod";
import { DefaultJSON, VisionJSON } from "./format/json";
import { post_process } from "./post-process";
import { OpenAIQuery } from "./query/openai";
import { ChatTask } from "./task/chat";
import { VisionChatTask } from "./task/vision-chat";
import type {
	AidTaskOptions,
	AidTaskRunner,
	BaseChatMessage,
	BaseChatParam,
	FormatEngine,
	QueryEngine,
	TaskBuilder,
	VisionChatParam,
} from "./types";

const log = debug("aid");

export class Aid<TaskGoal, CaseParam, FormatPayload, QueryPayload> {
	protected qe: QueryEngine<QueryPayload>;
	protected fe: FormatEngine<FormatPayload, QueryPayload>;
	protected tb: TaskBuilder<TaskGoal, CaseParam, FormatPayload>;

	constructor(
		qe: QueryEngine<QueryPayload>,
		fe: FormatEngine<FormatPayload, QueryPayload>,
		tb: TaskBuilder<TaskGoal, CaseParam, FormatPayload>,
	) {
		this.qe = qe;
		this.fe = fe;
		this.tb = tb;
	}

	task<Out>(
		goal: TaskGoal,
		expected: z.ZodType<Out>,
		opt?: AidTaskOptions<CaseParam, Out>,
	): AidTaskRunner<CaseParam, Out> {
		return async (input = opt?.default) => {
			const fp = await this.tb(goal, input, opt?.examples ?? []);
			log("fp", fp);

			const qp = await this.fe(fp, expected);
			log("qp", qp);

			const output = await this.qe(qp);
			log("output", output);

			const post = post_process(output);
			log("post-processed", post);

			const json = JSON.parse(post);

			if (!opt?.check) {
				return { result: json, errors: [] };
			}

			const result = expected.safeParse(json);
			if (result.success) {
				return { result: json, errors: [] };
			} else {
				return { result: json, errors: result.error.issues };
			}
		};
	}

	static from(
		openai: OpenAI,
		param: Omit<OpenAI.Chat.ChatCompletionCreateParamsNonStreaming, "messages">,
	): Aid<string, BaseChatParam, BaseChatMessage[], BaseChatMessage[]> {
		const qe = OpenAIQuery(openai, param);
		const fe = DefaultJSON({ strategy: "json-schema" });
		const tb = ChatTask();
		return new Aid(qe, fe, tb);
	}

	static chat(
		qe: QueryEngine<BaseChatMessage[]>,
		{
			tb = ChatTask(),
			fe = DefaultJSON({ strategy: "json-schema" }),
		}: {
			tb?: TaskBuilder<string, BaseChatParam, BaseChatMessage[]>;
			fe?: FormatEngine<BaseChatMessage[], BaseChatMessage[]>;
		} = {},
	): Aid<string, BaseChatParam, BaseChatMessage[], BaseChatMessage[]> {
		return new Aid(qe, fe, tb);
	}

	static vision(
		qe: QueryEngine<OpenAI.Chat.ChatCompletionMessageParam[]>,
		{
			tb = VisionChatTask(),
			fe = VisionJSON({ strategy: "json-schema" }),
		}: {
			tb?: TaskBuilder<string, VisionChatParam, OpenAI.Chat.ChatCompletionMessageParam[]>;
			fe?: FormatEngine<
				OpenAI.Chat.ChatCompletionMessageParam[],
				OpenAI.Chat.ChatCompletionMessageParam[]
			>;
		} = {},
	): Aid<
		string,
		VisionChatParam,
		OpenAI.Chat.ChatCompletionMessageParam[],
		OpenAI.Chat.ChatCompletionMessageParam[]
	> {
		return new Aid(qe, fe, tb);
	}
}
