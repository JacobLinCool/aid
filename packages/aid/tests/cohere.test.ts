import debug from "debug";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { Aid, BaseChatMessage, QueryEngine } from "../src";
import { WIKI_AI } from "./fixtures/articles";

debug.enable("aid");

describe("Cohere", () => {
	it(
		"Custom Model",
		async (c) => {
			const api_token = process.env.COHERE_API_TOKEN;
			if (!api_token) {
				c.skip();
				return;
			}

			const q: QueryEngine<BaseChatMessage[]> = async (messages) => {
				const message = messages[messages.length - 1].content;
				const chat_history = messages.slice(0, -1).map((message) => ({
					role: message.role === "user" ? "User" : "Chatbot",
					message: message.content,
				}));
				const res = await fetch("https://api.cohere.ai/v1/chat", {
					method: "POST",
					headers: {
						Authorization: `Bearer ${api_token}`,
						"Content-Type": "application/json",
					},
					body: JSON.stringify({
						model: "command",
						temperature: 0,
						connectors: [],
						documents: [],
						prompt_truncation: "auto",
						message,
						chat_history,
					}),
				});

				const result = (await res.json()) as { text: string };
				console.log(result);
				return result.text;
			};

			const aid = Aid.chat(q);

			const analyze = aid.task(
				"Summarize and extract keywords",
				z.object({
					summary: z.string().max(300),
					keywords: z.array(z.string().max(30)).max(10),
				}),
			);

			const { result } = await analyze(WIKI_AI);
			expect(result).toMatchSnapshot();
			expect(typeof result).toBe("object");
			expect(typeof result.summary).toBe("string");
			expect(Array.isArray(result.keywords)).toBe(true);
			result.keywords.forEach((keyword) => expect(typeof keyword).toBe("string"));
		},
		{ timeout: 60_000 },
	);
});
