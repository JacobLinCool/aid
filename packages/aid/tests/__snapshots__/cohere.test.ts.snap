// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cohere > Custom Model 1`] = `
{
  "keywords": [
    "AI",
    "machine intelligence",
    "deep learning",
    "artificial neural networks",
    "natural language processing",
    "optimization",
    "planning",
    "logic",
    "robotics",
    "computer science",
    "industry applications",
  ],
  "summary": "Artificial intelligence (AI) is the study of intelligent machines and their applications, which aim to mimic human intelligence in problem-solving and decision-making. AI technologies are widely used in various fields, including industry, government, and science, for tasks such as web search, recommendations, natural language processing, self-driving cars, and creative applications. While AI has faced challenges and setbacks since its establishment as an academic discipline in 1956, advancements in deep learning have spurred increased interest and funding in recent years.",
}
`;
